crates/ruff_linter/src/noqa.rs

```diff
@@ -418,182 +300,408 @@ impl<'a> FileNoqaDirectives<'a> {
 }
 
-/// An individual file-level exemption (e.g., `# ruff: noqa` or `# ruff: noqa: F401, F841`). Like
-/// [`FileNoqaDirectives`], but only for a single line, as opposed to an aggregated set of exemptions
-/// across a source file.
+/// Output of lexing a `noqa` directive.
 #[derive(Debug)]
-pub(crate) enum ParsedFileExemption<'a> {
-    /// The file-level exemption ignores all rules (e.g., `# ruff: noqa`).
-    All,
-    /// The file-level exemption ignores specific rules (e.g., `# ruff: noqa: F401, F841`).
-    Codes(Codes<'a>),
+pub(crate) struct NoqaLexerOutput<'a> {
+    warnings: Vec<LexicalWarning>,
+    directive: Directive<'a>,
 }
 
-impl<'a> ParsedFileExemption<'a> {
-    /// Return a [`ParsedFileExemption`] for a given `comment_range` in `source`.
-    fn try_extract(comment_range: TextRange, source: &'a str) -> Result<Option<Self>, ParseError> {
-        let line = &source[comment_range];
-        let offset = comment_range.start();
-        let init_line_len = line.text_len();
+/// Lexes in-line `noqa` comment, e.g. `# noqa: F401`
+fn lex_inline_noqa(
+    comment_range: TextRange,
+    source: &str,
+) -> Result<Option<NoqaLexerOutput<'_>>, LexicalError> {
+    let source = &source[..comment_range.end().to_usize()];
+    let offset = comment_range.start();
+    let lexer = NoqaLexer::starts_at(offset, source);
+    lexer.lex_inline_noqa()
+}
 
-        let line = Self::lex_whitespace(line);
-        let Some(line) = Self::lex_char(line, '#') else {
-            return Ok(None);
-        };
-        let comment_start = init_line_len - line.text_len() - '#'.text_len();
-        let line = Self::lex_whitespace(line);
+/// Lexes file-level exemption comment, e.g. `# ruff: noqa: F401`
+fn lex_file_exemption(
+    comment_range: TextRange,
+    source: &str,
+) -> Result<Option<NoqaLexerOutput<'_>>, LexicalError> {
+    let source = &source[..comment_range.end().to_usize()];
+    let offset = comment_range.start();
+    let lexer = NoqaLexer::starts_at(offset, source);
+    lexer.lex_file_exemption()
+}
 
-        let Some(line) = Self::lex_flake8(line).or_else(|| Self::lex_ruff(line)) else {
-            return Ok(None);
-        };
+pub(crate) fn lex_codes(text: &str) -> Result<Vec<Code<'_>>, LexicalError> {
+    let mut lexer = NoqaLexer::starts_at(TextSize::new(0), text);
+    lexer.lex_codes()?;
+    Ok(lexer.codes)
+}
 
-        let line = Self::lex_whitespace(line);
-        let Some(line) = Self::lex_char(line, ':') else {
-            return Ok(None);
-        };
-        let line = Self::lex_whitespace(line);
-        let Some(line) = Self::lex_noqa(line) else {
-            return Ok(None);
-        };
-        let line = Self::lex_whitespace(line);
+/// Lexer for `noqa` comment lines.
+///
+/// Assumed to be initialized with range or offset starting
+/// at the `#` at the beginning of a `noqa` comment.
+#[derive(Debug)]
+struct NoqaLexer<'a> {
+    /// A slice of source text starting at the beginning of a `noqa` comment
```

Comment:
Is this comment still accurate or is it the beginning of a comment (`#`)? We also don't know if it is a noqa comment at this point.